File: packages/contracts/src/sources.ts
import { z } from "zod";

export const Channel = z.enum(["email", "sms", "chat", "forum", "other"]);
export type Channel = z.infer<typeof Channel>;

export const CHANNEL_LABELS: Record<Channel, string> = {
  email: "Email",
  sms: "Text messages",
  chat: "Chat",
  forum: "Forums and threads",
  other: "Other",
};

export const ConnectorInfo = z.object({
  connector: z.string(),
  displayName: z.string(),
  channel: z.string(),
  description: z.string(),
  locationKind: z.enum(["file", "folder"]),
  extensions: z.array(z.string()),
});
export type ConnectorInfo = z.infer<typeof ConnectorInfo>;

export const SourceStatus = z.enum(["new", "ready", "importing", "imported", "failed"]);
export type SourceStatus = z.infer<typeof SourceStatus>;

export const Source = z.object({
  id: z.string(),
  connector: z.string(),
  name: z.string(),
  channel: z.string(),
  location: z.string().nullable(),
  config: z.record(z.string(), z.unknown()),
  status: z.string(),
  createdAt: z.string(),
  lastImportedAt: z.string().nullable(),
  messageCount: z.number(),
  lastError: z.object({ message: z.string() }).passthrough().nullable(),
});
export type Source = z.infer<typeof Source>;

/** What `validate` found, before the user commits to importing. */
export const ValidationReport = z.object({
  ok: z.boolean(),
  blockers: z.array(z.string()),
  warnings: z.array(z.string()),
  conversations: z.number(),
  messages: z.number(),
  frequentIdentifiers: z.array(z.tuple([z.string(), z.number()])),
  earliest: z.string().nullable(),
  latest: z.string().nullable(),
});
export type ValidationReport = z.infer<typeof ValidationReport>;

export const ImportSummary = z.object({
  conversations: z.number(),
  inserted: z.number(),
  duplicates: z.number(),
  empty: z.number(),
  fromSelf: z.number(),
  unattributed: z.number(),
  participantsCreated: z.number(),
});
export type ImportSummary = z.infer<typeof ImportSummary>;

/**
 * One line summarising an import, for the job tray. Says what was skipped and
 * why, because "45 messages imported" hides the 1 that could not be attributed.
 */
export function describeImport(s: ImportSummary): string {
  const parts = [`${s.inserted} messages from ${s.conversations} conversations`];
  if (s.fromSelf > 0) parts.push(`${s.fromSelf} written by you`);
  if (s.duplicates > 0) parts.push(`${s.duplicates} already imported`);
  if (s.unattributed > 0) parts.push(`${s.unattributed} with no identifiable author`);
  if (s.empty > 0) parts.push(`${s.empty} with no text`);
  return parts.join(", ");
}

/** How a mailbox signs in: an app password, or a sign-in with Microsoft. */
export const MailboxAuth = z.enum(["password", "microsoft"]);
export type MailboxAuth = z.infer<typeof MailboxAuth>;

/** How to reach a mailbox. The password is sent separately and never stored here. */
export const ImapAccount = z.object({
  host: z.string(),
  port: z.number(),
  username: z.string(),
  security: z.enum(["tls", "plain"]),
  /** Absent for a mailbox connected before 0.10.0-alpha.14: a password. */
  auth: MailboxAuth.optional(),
});
export type ImapAccount = z.infer<typeof ImapAccount>;

/** How a connected mailbox signs in, from its stored settings. */
export function mailboxAuth(source: {
  connector: string;
  config: Record<string, unknown>;
}): MailboxAuth | null {
  if (source.connector !== "imap") return null;
  const parsed = z
    .object({ account: z.object({ auth: MailboxAuth.optional() }) })
    .safeParse(source.config);
  return parsed.success ? (parsed.data.account.auth ?? "password") : "password";
}

/** What connecting found, before anything is read. */
export const ImapProbe = z.object({
  folders: z.array(z.string()),
  sentFolder: z.string().nullable(),
  counts: z.array(z.tuple([z.string(), z.number()])),
  warnings: z.array(z.string()),
});
export type ImapProbe = z.infer<typeof ImapProbe>;

/**
 * Server settings for the providers most people use. Most want an app
 * password for IMAP, not the account password; the dialog says so. Microsoft
 * takes no password at all: those sign in with Microsoft (`signIn`), when
 * this copy of Mimic was built able to.
 */
const MICROSOFT =
  "Outlook.com and Hotmail no longer let apps like this sign in with a password. You sign in with Microsoft in your browser instead.";

/** Said when this copy of Mimic can't sign in with Microsoft. */
export const MICROSOFT_UNAVAILABLE =
  "Outlook.com and Hotmail no longer let apps like this sign in with a password, and this copy of Mimic isn't set up to sign in with Microsoft. For now, bring your mail in as an .mbox file: add the account to Thunderbird, then export the inbox and sent folders as mbox.";

export const KNOWN_MAIL_HOSTS: Record<
  string,
  { host: string; port: number; help: string; signIn?: "microsoft" }
> = {
  "gmail.com": {
    host: "imap.gmail.com",
    port: 993,
    help: "Google Account → Security → 2-Step Verification → App passwords.",
  },
  "googlemail.com": {
    host: "imap.gmail.com",
    port: 993,
    help: "Google Account → Security → 2-Step Verification → App passwords.",
  },
  // Microsoft stopped accepting passwords of any kind for IMAP on personal
  // accounts in September 2024; these sign in with Microsoft instead.
  "outlook.com": { host: "outlook.office365.com", port: 993, help: MICROSOFT, signIn: "microsoft" },
  "hotmail.com": { host: "outlook.office365.com", port: 993, help: MICROSOFT, signIn: "microsoft" },
  "live.com": { host: "outlook.office365.com", port: 993, help: MICROSOFT, signIn: "microsoft" },
  "msn.com": { host: "outlook.office365.com", port: 993, help: MICROSOFT, signIn: "microsoft" },
  "icloud.com": {
    host: "imap.mail.me.com",
    port: 993,
    help: "appleid.apple.com → Sign-In and Security → App-Specific Passwords.",
  },
  "me.com": {
    host: "imap.mail.me.com",
    port: 993,
    help: "appleid.apple.com → Sign-In and Security → App-Specific Passwords.",
  },
  "fastmail.com": {
    host: "imap.fastmail.com",
    port: 993,
    help: "Fastmail → Settings → Privacy & Security → App passwords.",
  },
  "yahoo.com": {
    host: "imap.mail.yahoo.com",
    port: 993,
    help: "Yahoo Account security → Generate app password.",
  },
};

/**
 * Microsoft's own mail domains in other countries, which sign in as
 * outlook.com does. Named one by one rather than matched by shape: a shape
 * would take in live.io or outlook.co, which are other people's, and leave
 * someone there with no way to give a password.
 */
const MICROSOFT_ELSEWHERE = new Set(
  [
    "hotmail.co.uk hotmail.fr hotmail.de hotmail.it hotmail.es hotmail.be hotmail.ca hotmail.nl hotmail.se",
    "hotmail.co.jp hotmail.com.au hotmail.com.br hotmail.com.ar",
    "live.co.uk live.fr live.de live.it live.nl live.be live.ca live.at live.dk live.se live.no live.ie",
    "live.cl live.jp live.co.za live.com.au live.com.mx live.com.ar",
    "outlook.fr outlook.de outlook.it outlook.es outlook.be outlook.ie outlook.at outlook.dk outlook.pt",
    "outlook.cz outlook.sk outlook.hu outlook.cl outlook.jp outlook.kr outlook.sg outlook.ph outlook.my",
    "outlook.co.id outlook.co.il outlook.co.th outlook.com.au outlook.com.br outlook.com.ar outlook.com.gr",
    "outlook.com.tr outlook.com.vn windowslive.com passport.com",
  ].flatMap((line) => line.split(" ")),
);

/** Server settings guessed from an address, or null when the domain is unknown. */
export function guessMailHost(address: string) {
  const domain = address.split("@")[1]?.trim().toLowerCase();
  if (!domain) return null;
  return (
    KNOWN_MAIL_HOSTS[domain] ??
    (MICROSOFT_ELSEWHERE.has(domain) ? (KNOWN_MAIL_HOSTS["outlook.com"] ?? null) : null)
  );
}

File: apps/desktop/src-tauri/src/commands/sources.rs
//! Source connectors: what Mimic can read, and importing from it.

use std::sync::atomic::Ordering;

use mimic_core::sources::imap::{self, Auth, Credential, ImapAccount, ImapProbe, Security};
use mimic_core::sources::oauth::{self, OAuthError};
use serde::Serialize;
use serde_json::json;
use tauri::State;

use crate::error::{CommandError, CommandResult};
use crate::SharedState;

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ConnectorInfo {
    pub connector: String,
    pub display_name: String,
    pub channel: String,
    pub description: String,
    pub location_kind: String,
    pub extensions: Vec<String>,
}

#[tauri::command]
pub async fn list_connectors() -> CommandResult<Vec<ConnectorInfo>> {
    Ok(mimic_core::sources::all()
        .into_iter()
        .map(|s| {
            let m = s.metadata();
            ConnectorInfo {
                connector: m.connector.into(),
                display_name: m.display_name.into(),
                channel: m.channel.into(),
                description: m.description.into(),
                location_kind: match m.location_kind {
                    mimic_core::sources::LocationKind::File => "file".into(),
                    mimic_core::sources::LocationKind::Folder => "folder".into(),
                },
                extensions: m.extensions.iter().map(|e| e.to_string()).collect(),
            }
        })
        .collect())
}

#[tauri::command]
pub async fn list_sources(state: State<'_, SharedState>) -> CommandResult<Vec<mimic_core::db::Source>> {
    Ok(state.db.list_sources()?)
}

/// Read a file without importing it, so the user can see what Mimic found
/// before committing.
#[tauri::command]
pub async fn validate_source_file(
    connector: String,
    location: String,
) -> CommandResult<mimic_core::sources::ValidationReport> {
    let source = mimic_core::sources::by_connector(&connector)?;
    Ok(source.validate(std::path::Path::new(&location))?)
}

#[tauri::command]
pub async fn create_source(
    state: State<'_, SharedState>,
    connector: String,
    name: String,
    channel: String,
    location: Option<String>,
) -> CommandResult<mimic_core::db::Source> {
    // Fail here rather than at import time, when the user has walked away.
    mimic_core::sources::by_connector(&connector)?;
    let source = state.db.create_source(&mimic_core::db::NewSource {
        connector,
        name,
        channel,
        location,
        config: serde_json::Value::Null,
    })?;
    state.db.set_source_status(&source.id, "ready", None)?;
    Ok(state.db.get_source(&source.id)?.unwrap_or(source))
}

#[tauri::command]
pub async fn start_source_import(
    state: State<'_, SharedState>,
    source_id: String,
) -> CommandResult<mimic_core::db::Job> {
    if state.db.user_identifier_set()?.is_empty() {
        return Err(CommandError::new(
            "no_identity",
            "Tell Mimic which addresses are yours first — otherwise every message imports as 'unknown' and none of it can teach it how you write.",
        ));
    }
    let Some(source) = state.db.get_source(&source_id)? else {
        return Err(CommandError::new("not_found", "That source no longer exists."));
    };
    // A connected mailbox is not read from a file; "import again" means
    // "check it now".
    if source.connector == mimic_core::sources::imap::CONNECTOR {
        return Ok(state.jobs.enqueue(mimic_core::sources::imap::JOB_KIND, json!({ "sourceId": source_id }))?);
    }
    Ok(state.jobs.enqueue(mimic_core::import::JOB_KIND, json!({ "sourceId": source_id }))?)
}

fn mailbox_error(e: mimic_core::sources::imap::ImapError) -> CommandError {
    CommandError::new("mailbox", e.to_string())
}

/// Log in and look around without importing anything: which folders would
/// be read, how much is in them, and anything worth knowing first (no sent
/// folder, a mailbox bigger than the first check reads).
#[tauri::command]
pub async fn probe_mailbox(
    mut account: mimic_core::sources::imap::ImapAccount,
    password: String,
) -> CommandResult<mimic_core::sources::imap::ImapProbe> {
    account.auth = Auth::Password;
    tauri::async_runtime::spawn_blocking(move || imap::probe(&account, &Credential::Password(password)))
        .await
        .map_err(|e| CommandError::new("internal", e.to_string()))?
        .map_err(mailbox_error)
}

/// Connect a mailbox: check the login works, store the password in the
/// secret store (never the database), create the source, and start the
/// first check.
#[tauri::command]
pub async fn connect_mailbox(
    state: State<'_, SharedState>,
    mut account: mimic_core::sources::imap::ImapAccount,
    password: String,
    is_mine: bool,
) -> CommandResult<mimic_core::db::Source> {
    refuse_without_identity(&state)?;
    // What is kept is a password, whatever the page said.
    account.auth = Auth::Password;
    let probe_account = account.clone();
    let probe_password = password.clone();
    let probe = tauri::async_runtime::spawn_blocking(move || {
        imap::probe(&probe_account, &Credential::Password(probe_password))
    })
    .await
    .map_err(|e| CommandError::new("internal", e.to_string()))?
    .map_err(mailbox_error)?;
    add_mailbox(&state, account, probe, is_mine, &password)
}

fn refuse_without_identity(state: &crate::AppState) -> CommandResult<()> {
    if state.db.user_identifier_set()?.is_empty() {
        return Err(CommandError::new(
            "no_identity",
            "Tell me which addresses are yours first, so I can tell your mail from everyone else's.",
        ));
    }
    Ok(())
}

/// A mailbox that has been reached: its address becomes the user's when they
/// said it is theirs, its secret — the password, or the refresh token of a
/// sign-in — goes to the secret store (never the database), and the first
/// check starts.
fn add_mailbox(
    state: &crate::AppState,
    account: ImapAccount,
    probe: ImapProbe,
    is_mine: bool,
    secret: &str,
) -> CommandResult<mimic_core::db::Source> {
    // Only when the user said this mailbox is theirs rather than shared: its
    // address becomes one of theirs, so what is in Sent is read as their
    // writing. A shared mailbox (team@, support@) is left alone — what
    // colleagues sent from it is not the user's.
    if is_mine && account.username.contains('@') {
        state.db.add_user_identifier(mimic_core::db::IdentifierKind::Email, account.username.trim())?;
        // Mail from it already read, from an export of the same mailbox, is
        // folded back if whoever it was filed under is nothing but the user;
        // anyone else is listed in Settings for the user to decide.
        crate::commands::people::reconcile_identity(&state.db, &state.jobs, "connect_mailbox");
    }
    let config = imap::ImapSourceConfig {
        account: account.clone(),
        folders: probe.folders,
        sent_folder: probe.sent_folder,
        marks: Default::default(),
        last_checked_at: None,
    };
    let source = state.db.create_source(&mimic_core::db::NewSource {
        connector: imap::CONNECTOR.into(),
        name: account.username.clone(),
        channel: "email".into(),
        location: None,
        config: serde_json::to_value(&config).map_err(|e| CommandError::new("internal", e.to_string()))?,
    })?;
    if let Err(e) = state.secrets.set(&imap::secret_key(&source.id), secret) {
        // Without it the mailbox can never be checked; do not leave a source
        // behind that looks connected and is not.
        let _ = state.db.delete_source_and_contents(&source.id);
        let what = if account.auth == Auth::Password { "the password" } else { "the sign-in" };
        return Err(CommandError::new("secrets", format!("I couldn't store {what}: {e}")));
    }
    state.db.set_source_status(&source.id, "ready", None)?;
    state.jobs.enqueue(imap::JOB_KIND, json!({ "sourceId": source.id }))?;
    Ok(state.db.get_source(&source.id)?.unwrap_or(source))
}

/// Whether this copy of Mimic can sign in to a mailbox with Microsoft
/// (Outlook.com, Hotmail, Microsoft 365): it needs Mimic's client id with
/// Microsoft, compiled in.
#[tauri::command]
pub async fn mail_sign_in_available() -> CommandResult<bool> {
    Ok(oauth::microsoft_client_id().is_some())
}

/// Where a Microsoft mailbox is, for the address signed in to.
fn microsoft_account(email: &str) -> ImapAccount {
    let provider = oauth::Provider::microsoft();
    ImapAccount {
        host: provider.imap_host.into(),
        port: provider.imap_port,
        username: email.trim().to_string(),
        security: Security::Tls,
        auth: Auth::Microsoft,
    }
}

/// A sign-in under way. The next can't start until every holder has let go:
/// the command that started it, which holds it until it returns — through
/// looking at the mailbox, not just signing in — and the wait on the
/// browser, which may outlive a command whose caller went away.
struct SigningIn(SharedState);

impl SigningIn {
    fn begin(state: &SharedState) -> CommandResult<std::sync::Arc<Self>> {
        if state.sign_in.active.swap(true, Ordering::SeqCst) {
            return Err(CommandError::new(
                "busy",
                "A sign-in is already waiting on your browser. Finish it there, or stop it, first.",
            ));
        }
        state.sign_in.cancel.store(false, Ordering::SeqCst);
        Ok(std::sync::Arc::new(SigningIn(state.clone())))
    }

    /// Whether `cancel_mail_sign_in` stopped this sign-in.
    fn stopped(&self) -> bool {
        self.0.sign_in.cancel.load(Ordering::SeqCst)
    }
}

impl Drop for SigningIn {
    fn drop(&mut self) {
        self.0.sign_in.active.store(false, Ordering::SeqCst);
    }
}

fn sign_in_error(e: OAuthError) -> CommandError {
    match e {
        OAuthError::Canceled => CommandError::new("canceled", e.to_string()),
        other => CommandError::new("sign_in", other.to_string()),
    }
}

/// Mimic's client id with Microsoft, or the refusal that says this copy
/// can't sign in.
fn microsoft_client() -> CommandResult<&'static str> {
    oauth::microsoft_client_id()
        .ok_or_else(|| CommandError::new("not_configured", OAuthError::NotConfigured("Microsoft").to_string()))
}

/// Sign in with Microsoft in the user's browser: stoppable with
/// `cancel_mail_sign_in`, and given up after five minutes.
async fn microsoft_sign_in(
    app: &tauri::AppHandle,
    signing_in: &std::sync::Arc<SigningIn>,
    client: &'static str,
    email: &str,
) -> CommandResult<oauth::Tokens> {
    let (app, waiting, hint) = (app.clone(), signing_in.clone(), email.to_string());
    tauri::async_runtime::spawn_blocking(move || {
        let open = |url: &str| {
            use tauri_plugin_opener::OpenerExt;
            app.opener().open_url(url, None::<&str>).map_err(|e| e.to_string())
        };
        let stop = || waiting.stopped();
        oauth::sign_in(&oauth::Provider::microsoft(), client, Some(&hint), &open, &stop)
    })
    .await
    .map_err(|e| CommandError::new("internal", e.to_string()))?
    .map_err(sign_in_error)
}

/// Look at a mailbox through a sign-in, as `probe_mailbox` does through a
/// password. A sign-in that works but is not this mailbox's — the user signed
/// in as someone else — is said so.
async fn probe_signed_in(account: ImapAccount, access_token: String) -> CommandResult<ImapProbe> {
    let email = account.username.clone();
    tauri::async_runtime::spawn_blocking(move || imap::probe(&account, &Credential::AccessToken(access_token)))
        .await
        .map_err(|e| CommandError::new("internal", e.to_string()))?
        .map_err(|e| match e {
            imap::ImapError::SignIn(_) => CommandError::new(
                "sign_in",
                format!(
                    "That sign-in doesn't open {email}'s mailbox. Sign in as {email}. If you did, IMAP may be turned off for this mailbox; for a work or school account, your admin can turn it on."
                ),
            ),
            other => mailbox_error(other),
        })
}

fn no_refresh_token() -> CommandError {
    CommandError::new(
        "sign_in",
        "Microsoft signed you in, but gave me no way to stay signed in, so I couldn't keep checking this mailbox.",
    )
}

/// Sign in to a mailbox with Microsoft in the browser, and look around it,
/// without importing anything. What the sign-in brought back is held in
/// memory until `connect_signed_in_mailbox`, and nowhere else.
#[tauri::command]
pub async fn sign_in_to_mailbox(
    app: tauri::AppHandle,
    state: State<'_, SharedState>,
    email: String,
) -> CommandResult<ImapProbe> {
    let email = email.trim().to_string();
    if !email.contains('@') {
        return Err(CommandError::new("invalid", "That doesn't look like an email address."));
    }
    let client = microsoft_client()?;
    let signing_in = SigningIn::begin(state.inner())?;
    *state.sign_in.pending.lock().unwrap_or_else(|p| p.into_inner()) = None;
    let tokens = microsoft_sign_in(&app, &signing_in, client, &email).await?;
    let refresh_token = tokens.refresh_token.clone().ok_or_else(no_refresh_token)?;
    let probe = probe_signed_in(microsoft_account(&email), tokens.access_token.clone()).await?;
    // Stopped while the mailbox was being looked at: the dialog is gone, so
    // nothing is held for it. Decided under the lock `cancel_mail_sign_in`
    // takes, so a stop can't land between the look and the keeping.
    let mut pending = state.sign_in.pending.lock().unwrap_or_else(|p| p.into_inner());
    if signing_in.stopped() {
        return Err(sign_in_error(OAuthError::Canceled));
    }
    *pending = Some(crate::app_state::PendingSignIn { email, refresh_token, probe: probe.clone() });
    Ok(probe)
}

/// Connect the mailbox just signed in to, as `connect_mailbox` connects one
/// with a password.
#[tauri::command]
pub async fn connect_signed_in_mailbox(
    state: State<'_, SharedState>,
    email: String,
    is_mine: bool,
) -> CommandResult<mimic_core::db::Source> {
    refuse_without_identity(&state)?;
    let pending = state.sign_in.pending.lock().unwrap_or_else(|p| p.into_inner()).take();
    let Some(signed_in) = pending.filter(|p| p.email.eq_ignore_ascii_case(email.trim())) else {
        return Err(CommandError::new("sign_in", "I no longer have that sign-in. Sign in again."));
    };
    add_mailbox(&state, microsoft_account(&signed_in.email), signed_in.probe, is_mine, &signed_in.refresh_token)
}

/// Stop a sign-in waiting on the browser, and forget one that is waiting to
/// be connected.
#[tauri::command]
pub async fn cancel_mail_sign_in(state: State<'_, SharedState>) -> CommandResult<()> {
    let mut pending = state.sign_in.pending.lock().unwrap_or_else(|p| p.into_inner());
    state.sign_in.cancel.store(true, Ordering::SeqCst);
    *pending = None;
    Ok(())
}

/// Sign a connected Microsoft mailbox in again — Microsoft asked for it, or
/// the saved sign-in can't be unlocked on this Windows account — without
/// removing it and everything read from it. The new sign-in is tried on the
/// mailbox first, and kept only if it opens it.
#[tauri::command]
pub async fn sign_in_mailbox_again(
    app: tauri::AppHandle,
    state: State<'_, SharedState>,
    source_id: String,
) -> CommandResult<mimic_core::db::Source> {
    let source = state
        .db
        .get_source(&source_id)?
        .filter(|s| s.connector == imap::CONNECTOR)
        .ok_or_else(|| CommandError::new("not_found", "That mailbox isn't connected any more."))?;
    let config: imap::ImapSourceConfig = serde_json::from_value(source.config.clone()).map_err(|_| {
        CommandError::new("mailbox", "I can't read this mailbox's settings; remove it and connect it again.")
    })?;
    if config.account.auth != Auth::Microsoft {
        return Err(CommandError::new("mailbox", "This mailbox uses a password. Give it a new one instead."));
    }
    let client = microsoft_client()?;
    let signing_in = SigningIn::begin(state.inner())?;
    let tokens = microsoft_sign_in(&app, &signing_in, client, &config.account.username).await?;
    let refresh_token = tokens.refresh_token.clone().ok_or_else(no_refresh_token)?;
    probe_signed_in(config.account.clone(), tokens.access_token.clone()).await?;
    // Stopped while the mailbox was being looked at: the old sign-in stays.
    if signing_in.stopped() {
        return Err(sign_in_error(OAuthError::Canceled));
    }
    let (waiting, id) = (state.inner().clone(), source_id.clone());
    tauri::async_runtime::spawn_blocking(move || {
        waiting.mail_credentials.settle(&id, || waiting.secrets.set(&imap::secret_key(&id), &refresh_token))
    })
    .await
    .map_err(|e| CommandError::new("internal", e.to_string()))?
    .map_err(|e| CommandError::new("secrets", format!("I couldn't keep the sign-in: {e}")))?;
    if source.status == "failed" {
        state.db.set_source_status(&source_id, "ready", None)?;
    }
    check_again(&state, &source_id)?;
    state
        .db
        .get_source(&source_id)?
        .ok_or_else(|| CommandError::new("not_found", "That mailbox isn't connected any more."))
}

/// Check a mailbox now — unless a check is already waiting to start, which
/// will read the new password or sign-in. One already running read the old
/// one, so another is queued behind it.
fn check_again(state: &crate::AppState, source_id: &str) -> CommandResult<()> {
    let queued =
        state.db.list_jobs(500, true)?.iter().any(|j| {
            j.kind == imap::JOB_KIND && j.status == "queued" && j.payload["sourceId"].as_str() == Some(source_id)
        });
    if !queued {
        state.jobs.enqueue(imap::JOB_KIND, json!({ "sourceId": source_id }))?;
    }
    Ok(())
}

/// Give a connected mailbox its password again — after the provider's app
/// password was replaced, or when the saved one cannot be unlocked on this
/// Windows account — without removing the mailbox and everything imported
/// through it. The login is tried first, and nothing is saved if it fails.
#[tauri::command]
pub async fn set_mailbox_password(
    state: State<'_, SharedState>,
    source_id: String,
    password: String,
) -> CommandResult<mimic_core::db::Source> {
    let (db, secrets, id) = (state.db.clone(), state.secrets.clone(), source_id.clone());
    tauri::async_runtime::spawn_blocking(move || {
        imap::replace_password(&db, &id, &password, |pw| {
            secrets.set(&imap::secret_key(&id), pw).map_err(|e| e.to_string())
        })
    })
    .await
    .map_err(|e| CommandError::new("internal", e.to_string()))?
    .map_err(|e| match e {
        imap::PasswordError::Login(e) => mailbox_error(e),
        other => CommandError::new("mailbox", other.to_string()),
    })?;
    check_again(&state, &source_id)?;
    state
        .db
        .get_source(&source_id)?
        .ok_or_else(|| CommandError::new("not_found", "That mailbox isn't connected any more."))
}

/// Remove a source and everything imported through it.
#[tauri::command]
pub async fn delete_source(
    state: State<'_, SharedState>,
    source_id: String,
) -> CommandResult<mimic_core::privacy::DeletionReport> {
    crate::commands::voice::stop_measuring(&state);
    let report = state.db.delete_source_and_contents(&source_id)?;
    // A disconnected mailbox's password, or sign-in, goes with it — after any
    // check still trading that sign-in for a token has saved what it was
    // handed, so nothing is left behind. The mail is already gone, so a
    // failure here does not undo that; it is logged, never with the password.
    let waiting = state.inner().clone();
    let removed = tauri::async_runtime::spawn_blocking(move || {
        waiting.mail_credentials.settle(&source_id, || waiting.secrets.remove(&imap::secret_key(&source_id)))
    })
    .await;
    match removed {
        Ok(Ok(())) => {}
        Ok(Err(e)) => {
            tracing::warn!(target: "secrets", error = %e, "a removed mailbox's password could not be removed")
        }
        Err(e) => tracing::warn!(target: "secrets", error = %e, "removing a removed mailbox's password stopped"),
    }
    Ok(report)
}

#[tauri::command]
pub async fn pick_source_file(
    app: tauri::AppHandle,
    title: Option<String>,
    extensions: Vec<String>,
) -> CommandResult<Option<String>> {
    use tauri_plugin_dialog::DialogExt;
    let (tx, rx) = tokio::sync::oneshot::channel();
    let exts: Vec<&str> = extensions.iter().map(String::as_str).collect();
    app.dialog()
        .file()
        .set_title(title.unwrap_or_else(|| "Choose a file".into()))
        .add_filter("Supported exports", &exts)
        .pick_file(move |p| {
            let _ = tx.send(p.map(|f| f.to_string()));
        });
    Ok(rx.await.unwrap_or(None))
}
